File: microservice_matter/device_mapper.py
from typing import Dict, List, Any, Optional

MATTER_TO_COMMON_TYPES = {
    0x0100: "light",
    0x0101: "dimmer_light",
    0x010C: "color_light",
    0x010D: "extended_color_light",
    0x0302: "temperature_sensor",
    0x0015: "contact_sensor",
    0x0107: "occupancy_sensor",
    0x010A: "door_lock",
    0x0202: "thermostat",
    0x0303: "window_covering",
    0x0840: "pressure_sensor",
    0x0405: "humidity_sensor",
}

CLUSTER_TO_ATTRIBUTES = {
    0x0006: ["state"],
    0x0008: ["brightness"],
    0x0300: ["color_temp", "hue", "saturation"],
    0x0402: ["temperature"],
    0x0045: ["contact"],
    0x0406: ["occupancy"],
    0x0101: ["lock_state"],
    0x0201: ["target_temperature", "current_temperature", "mode"],
    0x0102: ["position"],
    0x0403: ["pressure"],
    0x0405: ["humidity"],
}

class DeviceMapper:
    @staticmethod
    def map_matter_device(matter_node: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        node_id = matter_node.get("node_id")
        device_type = matter_node.get("device_type_id")

        if device_type not in MATTER_TO_COMMON_TYPES:
            return None

        common_type = MATTER_TO_COMMON_TYPES[device_type]
        attributes = DeviceMapper._extract_attributes(matter_node)
        commands = DeviceMapper._get_available_commands(device_type)

        return {
            "id": f"matter_{node_id}",
            "type": common_type,
            "name": matter_node.get("name", f"Matter Device {node_id}"),
            "protocol": "matter",
            "attributes": attributes,
            "commands": commands,
            "metadata": {
                "node_id": node_id,
                "vendor_id": matter_node.get("vendor_id"),
                "product_id": matter_node.get("product_id"),
                "device_type_id": device_type,
            }
        }

    @staticmethod
    def _extract_attributes(matter_node: Dict[str, Any]) -> List[Dict[str, Any]]:
        attributes = []
        clusters = matter_node.get("clusters", [])

        for cluster in clusters:
            cluster_id = cluster.get("cluster_id")
            if cluster_id in CLUSTER_TO_ATTRIBUTES:
                for attr_name in CLUSTER_TO_ATTRIBUTES[cluster_id]:
                    attr_value = cluster.get("attributes", {}).get(attr_name)
                    if attr_value is not None:
                        attributes.append({
                            "name": attr_name,
                            "value": attr_value,
                            "type": DeviceMapper._get_attribute_type(attr_name)
                        })

        return attributes

    @staticmethod
    def _get_attribute_type(attr_name: str) -> str:
        type_mapping = {
            "state": "boolean",
            "brightness": "percentage",
            "color_temp": "color_temperature",
            "hue": "number",
            "saturation": "number",
            "temperature": "temperature",
            "contact": "boolean",
            "occupancy": "boolean",
            "lock_state": "boolean",
            "target_temperature": "temperature",
            "current_temperature": "temperature",
            "mode": "string",
            "position": "percentage",
            "pressure": "number",
            "humidity": "percentage",
        }
        return type_mapping.get(attr_name, "string")

    @staticmethod
    def _get_available_commands(device_type: int) -> List[str]:
        command_mapping = {
            0x0100: ["turn_on", "turn_off", "toggle"],
            0x0101: ["turn_on", "turn_off", "set_brightness"],
            0x010C: ["turn_on", "turn_off", "set_brightness", "set_color_temp"],
            0x010D: ["turn_on", "turn_off", "set_brightness", "set_color", "set_color_temp"],
            0x010A: ["lock", "unlock"],
            0x0202: ["set_temperature", "set_mode"],
            0x0303: ["open", "close", "set_position"],
        }
        return command_mapping.get(device_type, [])

File: microservice_zigbee/deviceMapper.js
const ZIGBEE_TO_COMMON_TYPES = {
  'light': 'light',
  'switch': 'switch',
  'dimmer': 'dimmer_light',
  'color_light': 'color_light',
  'temperature': 'temperature_sensor',
  'humidity': 'humidity_sensor',
  'contact': 'contact_sensor',
  'motion': 'motion_sensor',
  'occupancy': 'occupancy_sensor',
  'door_lock': 'door_lock',
  'plug': 'plug',
  'thermostat': 'thermostat',
  'cover': 'window_covering',
  'button': 'button',
  'vibration': 'vibration_sensor',
  'leak': 'leak_sensor',
  'smoke': 'smoke_sensor',
};

class DeviceMapper {
  static mapZigbeeDevice(zigbeeDevice) {
    const ieee = zigbeeDevice.ieee_address;
    const type = zigbeeDevice.definition?.type || 'unknown';

    if (!ZIGBEE_TO_COMMON_TYPES[type]) {
      return null;
    }

    const commonType = ZIGBEE_TO_COMMON_TYPES[type];
    const attributes = this._extractAttributes(zigbeeDevice);
    const commands = this._getAvailableCommands(type);

    return {
      id: `zigbee_${ieee}`,
      type: commonType,
      name: zigbeeDevice.friendly_name || `Zigbee Device ${ieee}`,
      protocol: 'zigbee',
      attributes,
      commands,
      metadata: {
        ieee_address: ieee,
        network_address: zigbeeDevice.network_address,
        model: zigbeeDevice.definition?.model,
        vendor: zigbeeDevice.definition?.vendor,
        power_source: zigbeeDevice.power_source,
        link_quality: zigbeeDevice.link_quality,
      }
    };
  }

  static _extractAttributes(zigbeeDevice) {
    const attributes = [];
    const state = zigbeeDevice.state || {};

    const attrMapping = {
      'state': { name: 'state', type: 'boolean', transform: (v) => v === 'ON' },
      'brightness': { name: 'brightness', type: 'percentage', transform: (v) => Math.round((v / 255) * 100) },
      'color_temp': { name: 'color_temp', type: 'color_temperature' },
      'color': { name: 'color', type: 'color' },
      'temperature': { name: 'temperature', type: 'temperature' },
      'humidity': { name: 'humidity', type: 'percentage' },
      'contact': { name: 'contact', type: 'boolean' },
      'occupancy': { name: 'occupancy', type: 'boolean' },
      'motion': { name: 'motion', type: 'boolean' },
      'lock_state': { name: 'lock_state', type: 'boolean' },
      'position': { name: 'position', type: 'percentage' },
      'voltage': { name: 'voltage', type: 'number' },
      'battery': { name: 'battery', type: 'percentage' },
      'power': { name: 'power', type: 'number' },
      'energy': { name: 'energy', type: 'number' },
      'current': { name: 'current', type: 'number' },
    };

    for (const [key, config] of Object.entries(attrMapping)) {
      if (state[key] !== undefined) {
        const value = config.transform ? config.transform(state[key]) : state[key];
        attributes.push({
          name: config.name,
          value,
          type: config.type
        });
      }
    }

    return attributes;
  }

  static _getAvailableCommands(type) {
    const commandMapping = {
      'light': ['turn_on', 'turn_off', 'toggle'],
      'switch': ['turn_on', 'turn_off', 'toggle'],
      'dimmer': ['turn_on', 'turn_off', 'set_brightness'],
      'color_light': ['turn_on', 'turn_off', 'set_brightness', 'set_color', 'set_color_temp'],
      'plug': ['turn_on', 'turn_off', 'toggle'],
      'door_lock': ['lock', 'unlock'],
      'thermostat': ['set_temperature', 'set_mode'],
      'cover': ['open', 'close', 'set_position'],
    };
    return commandMapping[type] || [];
  }

  static mapCommandToZigbee(command, params) {
    const commandMapping = {
      'turn_on': { state: 'ON' },
      'turn_off': { state: 'OFF' },
      'toggle': { state: 'TOGGLE' },
      'set_brightness': { brightness: Math.round((params.brightness / 100) * 255) },
      'set_color_temp': { color_temp: params.color_temp },
      'set_color': { color: params.color },
      'lock': { state: 'LOCK' },
      'unlock': { state: 'UNLOCK' },
      'set_temperature': { occupied_heating_setpoint: params.temperature },
      'set_mode': { system_mode: params.mode },
      'open': { state: 'OPEN' },
      'close': { state: 'CLOSE' },
      'set_position': { position: params.position },
    };

    return commandMapping[command] || null;
  }
}

module.exports = DeviceMapper;
